zkvm/src/predicate.rs

```diff
@@ -15,125 +16,151 @@ use crate::transcript::TranscriptProtocol;
 }
 
 impl Predicate {
-    pub fn opaque(point: CompressedRistretto) -> Self {
-        Predicate {
-            point,
-            witness: None,
+    /// Converts predicate to a compressed point
+    pub fn to_point(&self) -> CompressedRistretto {
+        match self {
+            Predicate::Opaque(p) => *p,
+            Predicate::Witness(w) => w.to_uncompressed_point().compress(),
         }
     }
 
-    /// Constructs a predicate from a witness.
-    pub fn from_witness(witness: PredicateWitness) -> Result<Self, VMError> {
-        Ok(Predicate {
-            point: witness.to_point()?,
-            witness: Some(witness),
-        })
+    /// Encodes the Predicate in program bytecode.
+    pub fn encode(&self, prog: &mut Vec<u8>) {
+        prog.extend_from_slice(&self.to_point().to_bytes())
     }
 
-    pub fn point(&self) -> CompressedRistretto {
-        self.point
+    /// Verifies whether the current predicate is a disjunction of two others.
+    /// Returns a `PointOp` instance that can be verified in a batch with other operations.
+    pub fn prove_or(&self, left: &Predicate, right: &Predicate) -> PointOp {
+        let l = left.to_point();
+        let r = right.to_point();
+        let f = Self::commit_or(l, r);
+
+        // P = L + f*B
+        PointOp {
+            primary: Some(f),
+            secondary: None,
+            arbitrary: vec![(Scalar::one(), l), (-Scalar::one(), self.to_point())],
+        }
     }
 
-    pub fn witness(&self) -> Option<&PredicateWitness> {
-        match &self.witness {
-            None => None,
-            Some(w) => Some(w),
+    /// Verifies whether the current predicate is a commitment to a program `prog`.
+    /// Returns a `PointOp` instance that can be verified in a batch with other operations.
+    pub fn prove_program_predicate(&self, prog: &[u8]) -> PointOp {
+        let h = Self::commit_program(prog);
+        // P == h*B2   ->   0 == -P + h*B2
+        PointOp {
+            primary: None,
+            secondary: Some(h),
+            arbitrary: vec![(-Scalar::one(), self.to_point())],
         }
     }
 
-    /// Encodes the Predicate in program bytecode.
-    pub fn encode(&self, prog: &mut Vec<u8>) {
-        prog.extend_from_slice(&self.point().to_bytes())
+    /// Creates an opaque predicate witness from a compressed point
+    pub fn opaque_witness(point: CompressedRistretto) -> Result<Self, VMError> {
+        Ok(Predicate::Witness(PredicateWitness::Opaque(
+            point.decompress().ok_or(VMError::FormatError)?,
+        )))
     }
 
-    /// Computes a disjunction of two predicates.
-    /// TBD: push this code into to_point() impl for the witness
-    pub fn or(self, right: Predicate) -> Result<Predicate, VMError> {
-        Ok(Predicate::from_witness(PredicateWitness::Or(
-            Box::new(self),
-            Box::new(right),
-        ))?)
+    /// Creates a predicate with a signing key witness
+    pub fn from_signing_key(secret_key: Scalar) -> Self {
```

Comment:
Similarly here, it might be more clearer to have the function name indicate that we're creating a `Witness`-type predicate (or to use `From<Scalar>` for `PredicateWitness` instead)

Reply:
My goal was to deemphasize the PredicateWitness in order to minimize amount of entities to think about (can't make it private since it's used in the public enum Predicate, but i removed all methods from it).

Reply:
renamed PredicateWitness::Opaque into PW::OpaqueBranch